File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-appservice-java/src/main/java/com/microsoft/azure/toolkit/intellij/legacy/appservice/AppServiceInfoBasicPanel.java
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */

package com.microsoft.azure.toolkit.intellij.legacy.appservice;

import com.intellij.openapi.project.Project;
import com.intellij.ui.TitledSeparator;
import com.microsoft.azure.toolkit.intellij.common.AzureArtifact;
import com.microsoft.azure.toolkit.intellij.common.AzureArtifactComboBox;
import com.microsoft.azure.toolkit.intellij.common.AzureFormPanel;
import com.microsoft.azure.toolkit.intellij.legacy.appservice.platform.RuntimeComboBox;
import com.microsoft.azure.toolkit.lib.Azure;
import com.microsoft.azure.toolkit.lib.appservice.config.AppServiceConfig;
import com.microsoft.azure.toolkit.lib.appservice.config.FunctionAppConfig;
import com.microsoft.azure.toolkit.lib.appservice.config.RuntimeConfig;
import com.microsoft.azure.toolkit.lib.appservice.model.Runtime;
import com.microsoft.azure.toolkit.lib.auth.Account;
import com.microsoft.azure.toolkit.lib.auth.AzureAccount;
import com.microsoft.azure.toolkit.lib.common.form.AzureFormInput;
import com.microsoft.azure.toolkit.lib.common.model.Subscription;
import lombok.SneakyThrows;
import org.apache.commons.compress.utils.FileNameUtils;
import org.apache.commons.lang3.ObjectUtils;
import org.apache.commons.lang3.StringUtils;

import javax.swing.*;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.function.Supplier;

import static com.microsoft.azuretools.utils.WebAppUtils.isSupportedArtifactType;

public class AppServiceInfoBasicPanel<T extends AppServiceConfig> extends JPanel implements AzureFormPanel<T> {
    private final Project project;
    private final Supplier<? extends T> supplier;
    private T config;

    private JPanel contentPanel;
    private AppNameInput textName;
    private RuntimeComboBox selectorRuntime;
    private AzureArtifactComboBox selectorApplication;
    private TitledSeparator deploymentTitle;
    private JLabel lblArtifact;
    private JLabel lblName;
    private JLabel lblPlatform;

    public AppServiceInfoBasicPanel(final Project project, final Supplier<? extends T> defaultConfigSupplier) {
        super();
        this.project = project;
        this.supplier = defaultConfigSupplier;
        $$$setupUI$$$(); // tell IntelliJ to call createUIComponents() here.
        this.init();
    }

    private void init() {
        this.textName.setRequired(true);
        this.selectorRuntime.setRequired(true);

        this.selectorApplication.setFileFilter(virtualFile -> {
            final String ext = FileNameUtils.getExtension(virtualFile.getPath());
            final Runtime platform = this.selectorRuntime.getValue();
            return StringUtils.isNotBlank(ext) && isSupportedArtifactType(platform, ext);
        });
        this.setDeploymentVisible(false);

        this.lblName.setLabelFor(textName);
        this.lblPlatform.setLabelFor(selectorRuntime);
        this.lblArtifact.setLabelFor(selectorApplication);
    }

    @SneakyThrows
    @Override
    public T getValue() {
        final String name = this.textName.getValue();
        final Runtime platform = this.selectorRuntime.getValue();
        final AzureArtifact artifact = this.selectorApplication.getValue();
        final Account account = Azure.az(AzureAccount.class).account();
        final Subscription subscription = Optional.ofNullable(config).map(AppServiceConfig::subscriptionId)
                                                  .map(account::getSubscription)
                                                  .orElseGet(() -> account.getSelectedSubscriptions().stream().findFirst().orElse(null));
        final T result = this.config == null ? supplier.get() : this.config;
        result.appName(name);
        Optional.ofNullable(platform).map(RuntimeConfig::fromRuntime).ifPresent(result::runtime);
        Optional.ofNullable(subscription).map(Subscription::getId).ifPresent(result::subscriptionId);
        // todo: web app creation dialog should use run/deploy configuration, which should be parent of info panel
//        if (Objects.nonNull(artifact)) {
//            final AzureArtifactManager manager = AzureArtifactManager.getInstance(this.project);
//            final String path = this.selectorApplication.getValue().getFileForDeployment();
//            result.setApplication(Paths.get(path));
//        }
        this.config = result;
        this.config.appSettings(ObjectUtils.firstNonNull(this.config.appSettings(), new HashMap<>()));
        return config;
    }

    @Override
    public void setValue(final T config) {
        this.config = config;
        final Subscription subscription = Optional.ofNullable(config.subscriptionId())
                                                  .map(Azure.az(AzureAccount.class).account()::getSubscription)
                                                  .orElse(null);
        this.textName.setValue(config.appName());
        this.textName.setSubscription(subscription);
        Optional.ofNullable(config.runtime())
                .map(c -> config instanceof FunctionAppConfig ? RuntimeConfig.toFunctionAppRuntime(c) : RuntimeConfig.toWebAppRuntime(c))
                .ifPresent(this.selectorRuntime::setValue);
    }

    @Override
    public List<AzureFormInput<?>> getInputs() {
        final AzureFormInput<?>[] inputs = {
            this.textName,
            this.selectorRuntime
        };
        return Arrays.asList(inputs);
    }

    @Override
    public void setVisible(final boolean visible) {
        this.contentPanel.setVisible(visible);
        super.setVisible(visible);
    }

    public RuntimeComboBox getSelectorRuntime() {
        return selectorRuntime;
    }

    private void createUIComponents() {
        // TODO: place custom component creation code here
        this.selectorApplication = new AzureArtifactComboBox(project, true);
    }

    public void setDeploymentVisible(boolean visible) {
        this.deploymentTitle.setVisible(visible);
        this.lblArtifact.setVisible(visible);
        this.selectorApplication.setVisible(visible);
    }

    public void setFixedRuntime(final Runtime runtime) {
        selectorRuntime.setPlatformList(Collections.singletonList(runtime));
    }
}

File: PluginsAndFeatures/azure-toolkit-for-rider/azure-intellij-plugin-appservice-dotnet/src/main/kotlin/com/microsoft/azure/toolkit/intellij/legacy/appservice/AppServiceInfoBasicPanel.kt
/*
 * Copyright 2018-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the MIT license.
 */

package com.microsoft.azure.toolkit.intellij.legacy.appservice

import com.intellij.openapi.Disposable
import com.intellij.openapi.util.Disposer
import com.intellij.ui.components.JBRadioButton
import com.intellij.ui.dsl.builder.ButtonsGroup
import com.intellij.ui.dsl.builder.Cell
import com.intellij.ui.dsl.builder.bind
import com.intellij.ui.dsl.builder.panel
import com.microsoft.azure.toolkit.intellij.common.AzureFormPanel
import com.microsoft.azure.toolkit.lib.Azure
import com.microsoft.azure.toolkit.lib.appservice.config.AppServiceConfig
import com.microsoft.azure.toolkit.lib.appservice.config.RuntimeConfig
import com.microsoft.azure.toolkit.lib.appservice.model.OperatingSystem
import com.microsoft.azure.toolkit.lib.appservice.model.Runtime
import com.microsoft.azure.toolkit.lib.auth.AzureAccount
import com.microsoft.azure.toolkit.lib.common.form.AzureFormInput
import java.util.function.Supplier
import javax.swing.JPanel

class AppServiceInfoBasicPanel<T>(
    private val defaultConfigSupplier: Supplier<T>
) : JPanel(), Disposable, AzureFormPanel<T> where T : AppServiceConfig {

    private var config: T?

    private val panel: JPanel
    private val textName = AppNameInput().apply {
        isRequired = true
    }

    private var operatingSystem: OperatingSystem
    private lateinit var operatingSystemGroup: ButtonsGroup
    private lateinit var windowsRadioButton: Cell<JBRadioButton>
    private lateinit var linuxRadioButton: Cell<JBRadioButton>
    private lateinit var dockerRadioButton: Cell<JBRadioButton>

    init {
        operatingSystem = OperatingSystem.WINDOWS

        Disposer.register(this, textName)

        panel = panel {
            group("Instance Details") {
                row("Name:") {
                    cell(textName)
                    label(".azurewebsites.net")
                }
                operatingSystemGroup = buttonsGroup {
                    row("Operating System:") {
                        windowsRadioButton = radioButton("Windows", OperatingSystem.WINDOWS)
                        linuxRadioButton = radioButton("Linux", OperatingSystem.LINUX)
                        dockerRadioButton = radioButton("Docker", OperatingSystem.DOCKER)
                            .visible(false)
                    }
                }.bind(::operatingSystem)
            }
        }

        add(panel)

        config = defaultConfigSupplier.get().also { setValue(it) }
    }

    override fun getValue(): T {
        val name = textName.value
        val operatingSystem =
            if (windowsRadioButton.component.isSelected) OperatingSystem.WINDOWS
            else if (linuxRadioButton.component.isSelected) OperatingSystem.LINUX
            else OperatingSystem.DOCKER

        val account = Azure.az(AzureAccount::class.java).account()
        val subscription = config
            ?.let { account.getSubscription(it.subscriptionId) }
            ?: account.selectedSubscriptions.firstOrNull()

        val result = config ?: defaultConfigSupplier.get()
        result.appName = name
        subscription?.let { result.subscriptionId = it.id }
        result.runtime = (result.runtime ?: RuntimeConfig()).apply {
            os = operatingSystem
        }

        config = result

        return result
    }

    override fun setValue(config: T) {
        this.config = config

        val subscription =
            config.subscriptionId?.let { Azure.az(AzureAccount::class.java).account().getSubscription(it) }

        textName.value = config.appName
        textName.setSubscription(subscription)
        when (config.runtime.os) {
            OperatingSystem.WINDOWS -> windowsRadioButton.component.isSelected = true
            OperatingSystem.LINUX -> linuxRadioButton.component.isSelected = true
            OperatingSystem.DOCKER -> dockerRadioButton.component.isSelected = true
            else -> linuxRadioButton.component.isSelected = true
        }
    }

    override fun getInputs(): List<AzureFormInput<*>> = listOf(textName)

    override fun setVisible(visible: Boolean) {
        panel.isVisible = visible
        super<JPanel>.setVisible(visible)
    }

    fun setFixedRuntime(runtime: Runtime) {
        when (runtime.operatingSystem) {
            OperatingSystem.WINDOWS -> {
                windowsRadioButton.component.isSelected = true
            }

            OperatingSystem.LINUX -> {
                linuxRadioButton.component.isSelected = true
            }

            OperatingSystem.DOCKER -> {
                dockerRadioButton.component.isSelected = true
            }

            else -> {
                linuxRadioButton.component.isSelected = true
            }
        }
        operatingSystemGroup.visible(false)
    }

    override fun dispose() {
    }
}